src/Common/LRUResourceCache.h

```diff
@@ -0,0 +1,255 @@
+        it->second.reference_count += 1;
+        queue.splice(queue.end(), queue, it->second.queue_iterator);
+        return it->second.value;
+    }
+
+    // mark a reference is released
+    void release(const Key & key)
+    {
+        std::lock_guard lock(mutex);
+        auto it = cells.find(key);
+        if (it == cells.end() || it->second.reference_count == 0)
+        {
+            LOG_ERROR(&Poco::Logger::get("LRUResourceCache"), "try to release an invalid element");
+            abort();
+        }
+        it->second.reference_count -= 1;
+    }
+
+    // If you want to update a value, call tryRemove() at first and then call acquire() with load_func.
+    bool tryRemove(const Key & key)
+    {
+        std::lock_guard guard(mutex);
+        auto it = cells.find(key);
+        if (it == cells.end())
+            return true;
+        auto & cell = it->second;
+        if (cell.reference_count)
+            return false;
+        queue.erase(cell.queue_iterator);
+        current_weight -= cell.weight;
+        cells.erase(it);
+        return true;
+    }
+
+    size_t weight()
+    {
+        std::lock_guard lock(mutex);
+        return current_weight;
+    }
+
+    size_t size()
+    {
+        std::lock_guard lock(mutex);
+        return cells.size();
+    }
+
+    void getStats(size_t & out_hits, size_t & out_misses, size_t & out_evict_count) const
+    {
+        out_hits = hits;
+        out_misses = misses;
+        out_evict_count = evict_count;
+    }
+
+private:
+    mutable std::mutex mutex;
+
+    using LRUQueue = std::list<Key>;
+    using LRUQueueIterator = typename LRUQueue::iterator;
+
+    struct Cell
+    {
+        MappedPtr value;
+        size_t weight = 0;
+        LRUQueueIterator queue_iterator;
+        size_t reference_count = 0;
+    };
+
+    using Cells = std::unordered_map<Key, Cell, HashFunction>;
+    Cells cells;
+    LRUQueue queue;
+    size_t current_weight = 0;
+    size_t max_weight = 0;
+    size_t max_element_size = 0;
+
+    struct InsertToken
+    {
+        std::mutex mutex;
+        MappedPtr value;
+        size_t reference_count = 0;
+    };
+    using InsertTokens = std::unordered_map<Key, InsertToken, HashFunction>;
+    InsertTokens insert_tokens;
+    WeightFunction weight_function;
+    std::atomic<size_t> hits{0};
+    std::atomic<size_t> misses{0};
+    std::atomic<size_t> evict_count{0};
+
+    InsertToken * acquireInsertToken(const Key & key)
+    {
+        auto & token = insert_tokens[key];
+        token.reference_count += 1;
+        return &token;
+    }
+
+    void releaseInsertToken(const Key & key)
+    {
+        auto it = insert_tokens.find(key);
+        if (it != insert_tokens.end())
+        {
+            it->second.reference_count -= 1;
+            if (it->second.reference_count == 0)
+                insert_tokens.erase(it);
+        }
+    }
+
+    // key mustn't be in the cache
+    Cell * insert_value(const Key & insert_key, MappedPtr value)
```

Comment:
Style. Should be `insertValue` at least. But I like name `set(...)` more (or `setImpl`).